### ISSUE TITLE
Race condition in _onFocus

### ISSUE THREAD

**Issue by brian-cruickshank** [PolymerElements/iron-menu-behavior#49]
In version 1.1.1, the _onFocus method contains the following code: 

```
      // clear the cached focus item
      this._defaultFocusAsync = this.async(function() {
        // focus the selected item when the menu receives focus, or the first item
        // if no item is selected
        var selectedItem = this.multi ? (this.selectedItems && this.selectedItems[0]) : this.selectedItem;
        this._setFocusedItem(null);
        if (selectedItem) {
          this._setFocusedItem(selectedItem);
        } else {
          this._setFocusedItem(this.items[0]);
        }
      // async 1ms to wait for `select` to get called from `_itemActivate`
      }, 1); 
```

The async 1ms wait used to be 100ms in version 1.0.2.   I'm finding that with v1.1.1, a delay of around 150ms is required for my application to work reliably with the debugger open.  Delays less than 100ms frequently cause the scenario outlined in [paper-menu/issues/17](https://github.com/PolymerElements/paper-menu/issues/17) to occur - i.e. after opening a submenu that has more items than can be fit within the available height and then clicking on the last menu item, the selection jumps up to the first menu item.  This occurs because selectedItem is null in the above code, which results in this._setFocusedItem(this.items[0]) to be called.

There are a number of factors which will impact the delay required to ensure that select is called before this async code executes, leading to race conditions that can be difficult to debug.  

At the very least, the delay should be increased to e.g. 150ms.  Ideally some form of event-driven callback mechanism would be preferred.

(See [paper-menu issue #89](https://github.com/PolymerElements/paper-menu/issues/89) for more info on additional step required to fix the menu jumping problem)

**Comment by bicknellr** [comment]
Hi @brian-cruickshank, I'm having trouble reproducing this issue. [This was my attempt](http://jsbin.com/miwoca/1/edit?html,output); I tried clicking on one of the earlier menu items, scrolling until it was out of view, and then clicking on a different menu item. As mentioned in [CONTRIBUTING.md](https://github.com/PolymerElements/iron-menu-behavior/blob/master/CONTRIBUTING.md), a reduced example would be very helpful. Could you provide a test case where this can be reproduced? ([Here's a template you can start from.](https://jsbin.com/xukike/edit?html,output)) Also, does this happen often enough to be pretty easily noticed?

**Comment by brian-cruickshank** [comment]
I tried to use the jsbin template you provided but can't get past the jsbin error "ReferenceError: HTMLImports is not defined... ", so I'm posting the code here.  It replicates the problem on Chrome 48.0.2564.116 m and Firefox 38.6.1.  I'm using Polymer 1.2.4.

Re: _does this happen often enough to be pretty easily noticed?_
Yes.  We're typically limited to half the height of the browser to display the menus in, so the top-level categories are always being expanded and then scrolled.   It's quite frustrating to users.

```
<!doctype html>
<html>
<head>
  <base href="http://polygit.org/components/">
  <script src="webcomponentsjs/webcomponents.js"></script>
  <link rel="import" href="polymer/polymer.html">
  <link rel="import" href="iron-icons/iron-icons.html">
<link rel="import" href="iron-icon/iron-icon.html">
<link rel="import" href="paper-item/paper-item.html">
<link rel="import" href="paper-menu/paper-menu.html">
<link rel="import" href="paper-menu/paper-submenu.html">
<link rel="import" href="iron-flex-layout/iron-flex-layout.html">
<link rel="import" href="iron-resizable-behavior/iron-resizable-behavior.html">
</head>
<body>


  <dom-module id="my-element">
     <template>
        <style is="custom-style" >
            [layout] {
            @apply(--layout);
            }
            [layout][vertical] {
            @apply(--layout-vertical);
            }
            :host[layout][horizontal] {
            @apply(--layout-horizontal);
            }
            [layout][flex] {
            @apply(--layout-flex);
            }
            .vertical {
            @apply(--layout-vertical);
            }
            .flex {
            @apply(--layout-flex);
            }
            #list {
                padding-top:10px;
            }

        </style>

            <template is="dom-if" restamp="true" if="true">
                <template is="dom-repeat" restamp="true" items="{{categories}}" as="category"  >
                    <paper-menu>
                        <paper-submenu>
                            <paper-item class="menu-trigger" layout horizontal>{{category.name}}<div flex class="flex item-content">&nbsp;</div>
                                <iron-icon class="item-content" icon="expand-more" ></iron-icon>
                            </paper-item>
                            <paper-menu class="menu-content sublist group-item">
                                <template is="dom-repeat" items="{{category.list}}" as="listItem">
                                    <div class="simple-item" tag="{{listItem.tag}}" title$="{{listItem.id}}" loaded$="{{listItem.importsLoaded}}" hidden$="{{listItem.isHidden}}" on-tap="menuItemClicked" on-contextmenu="showContextMenu">{{listItem.label}}</div>
                                </template>
                            </paper-menu>
                        </paper-submenu>
                    </paper-menu>
                </template>
            </template>
    </template>

    <script>
      HTMLImports.whenReady(function() {
        Polymer({
          is: 'my-element',
          properties: {
               categories: {
                    notify: true,                 
                    type: Array
                }
            },
            showContextMenu: function (event){
                console.log("showContextMenu clicked");
            },
            previousSelectedItem: undefined,
            menuItemClicked: function (event){
                var sender = event.target;
                console.log("menuItemClicked");
                    if (this.previousSelectedItem !== undefined){
                        this.previousSelectedItem.style.backgroundColor= 'white';
                    }
                    this.previousSelectedItem = sender;
                    sender.style.backgroundColor = "#eeeeee";
            },
            hostAttributes: {
                'vertical': '',
                'layout': ''
            },
            makeCategory: function(catName, itemName, numItems){
               var result = {name: catName,list:[]};
               for (var i=0; i<numItems; i++) {
                 result.list.push({
                   label:itemName+i,
                   tag:'tooltip'+itemName+i,
                   id:'id_'+itemName+i,
                   importsLoaded: true,
                   isHidden: false
                 })
               }
                return result;
            },
            ready: function(){
               this.categories = [];
               this.push('categories',this.makeCategory('Category A1','a1Item_',15));
               this.push('categories',this.makeCategory('Category B2','b2Item_',25));
               this.push('categories',this.makeCategory('Category C3','c3Item_',15));
               this.push('categories',this.makeCategory('Category D4','d4Item_',35));
            }
        });
      });
    </script>
  </dom-module>

  <my-element></my-element>
</body>
</html>
```

**Comment by ktiedt** [comment]
Hitting this too.
It is rather critical for us. (the other ticket #89 with a proposed fix appears to be getting no attention)

**Comment by bicknellr** [comment]
Closed in #51.
